include/psen_scan_v2/udp_client.h

```diff
@@ -135,6 +133,23 @@ class UdpClientImpl
 
   boost::asio::ip::udp::socket socket_;
   boost::asio::ip::udp::endpoint endpoint_;
+
+public:
+  class CloseConnectionFailure : public std::runtime_error
+  {
+  public:
+    CloseConnectionFailure(const std::string& msg = "Failure while closing connection") : std::runtime_error(msg)
+    {
+    }
+  };
+
+  class OpenConnectionFailure : public std::runtime_error
+  {
+  public:
+    OpenConnectionFailure(const std::string& msg = "Failure while opening connection") : std::runtime_error(msg)
+    {
+    }
+  };
 };
```

Comment:
Please, state all public stuff before the private stuff. This way the reader can easily overview the public interface of the class. Furthermore, ensure that first all public methods are stated and then all public members, etc.

I suggest that you state the possible exceptions at the beginning of the class. Please, separate the function declaration and the definition to ease readability.